lib/db/seeder.rb

```diff
@@ -0,0 +1,200 @@
+      [:replicate_from_nodes, :replicate_to_nodes,
+       :restore_from_nodes, :restore_to_nodes
+      ].each do |agreement_type|
+        create_node_agreements(node_entries, agreement_type)
+      end
+    end
+
+    # @api private
+    # @param node_entries [Array<Hash>] The hash elements should each have
+    #   the keys :namespace, :"#{agreement_type}"
+    # @param agreement_type [Symbol] The method name of the agreement type,
+    #   e.g. :replicate_to_nodes
+    def create_node_agreements(node_entries, agreement_type)
+      node_entries.each do |node_entry|
+        node, new_nodes, duplicate_nodes = discover_nodes(node_entry, agreement_type)
+
+        duplicate_nodes.each do |existing_node|
+          report_skip("#{agreement_type} for #{node_entry[:namespace]}", existing_node.namespace)
+        end
+
+        new_nodes.each do |new_node|
+          report("#{agreement_type} for #{node_entry[:namespace]}", new_node.namespace) do
+            node.public_send(agreement_type) << new_node
+          end
+        end
+      end
+    end
+
+    private
+
+    attr_reader :logger, :maps
+
+    def create_from_entries(entries, klass, display_block, convert_block)
+      entries.each do |entry|
+        report(klass, display_block.call(entry)) do
+          klass.create! convert_block.call(entry)
+        end
+      end
+    end
+
+    def discover_nodes(node_entry, agreement_type)
+      node = Node.find_by_namespace(node_entry[:namespace])
+      seed_nodes = Node.where(namespace: node_entry[agreement_type])
+      duplicate_nodes = seed_nodes & node.public_send(agreement_type)
+      new_nodes = seed_nodes - duplicate_nodes
+      return node, new_nodes, duplicate_nodes
+    end
+
+    DUPLICATE_ERROR_MESSAGE = ["has already been taken"].freeze
+    def should_skip_invalid?(errors)
+      errors.values.delete_if do |e|
+        e == DUPLICATE_ERROR_MESSAGE
+      end.empty?
+    end
+
+    def report_success(display_name, display_obj)
+      logger.info("Created #{display_name}: #{display_obj}")
+    end
+
+    def report_skip(display_name, display_obj)
+      logger.info("Skipped existing #{display_name}: #{display_obj}")
+    end
+
+    def report_fail(display_name, display_obj)
+      logger.error("Could not create #{display_name}: #{display_obj}")
+    end
+
+    def simple_record_from_entry(entry)
+      { name: entry }
+    end
+
+    def node_from_entry(node_entry)
+      {
+        namespace: node_entry[:namespace],
+        name: node_entry[:name],
+        storage_region: StorageRegion.find_by_name(node_entry[:storage_region]),
+        storage_type: StorageType.find_by_name(node_entry[:storage_type]),
+        fixity_algs: FixityAlg.where(name: node_entry[:fixity_algs]),
+        protocols: Protocol.where(name: node_entry[:protocols]),
+        api_root: node_entry[:api_root],
+        private_auth_token: node_entry[:private_auth_token],
+        auth_credential: node_entry[:auth_credential]
+      }
+    end
```

Comment:
BTW, also curious about whether the fabricator could be used in this class, maybe with a PORO generator (they have an example on their docs).